src/GraphQL/Execution/ExecutionStrategy.cs

```diff
@@ -357,15 +357,49 @@ void SetArrayItemNode(object d)
                         }
                         else if (node is ValueExecutionNode valueNode)
                         {
-                            node.Result = valueNode.GraphType.Serialize(d)
-                                ?? throw new InvalidOperationException($"Unable to serialize '{d}' to '{valueNode.GraphType.Name}' for list index {index}.");
+                            object serializedResult = null;
+
+                            try
+                            {
+                                serializedResult = valueNode.GraphType.Serialize(d);
+                            }
+                            catch (Exception ex)
+                            {
+                                // Stop execution as early as possible, it doesn't make sense to set the rest of the array elements and then execute nodes
+                                if (listType.ResolvedType is NonNullGraphType)
+                                {
+                                    throw new InvalidOperationException($"Unable to serialize '{d}' to '{valueNode.GraphType.Name}' for list index {index}.", ex);
+                                }
+                                else
+                                {
+                                    SetNodeError(context, node, new ExecutionError($"Unable to serialize '{d}' to '{valueNode.GraphType.Name}'.", ex));
+                                    arrayItems.Add(node);
+                                    return;
+                                }
+                            }
+
+                            node.Result = serializedResult;
+
+                            if (serializedResult == null)
+                            {
+                                // Stop execution as early as possible, it doesn't make sense to set the rest of the array elements and then execute nodes
+                                if (listType.ResolvedType is NonNullGraphType)
+                                {
+                                    throw new InvalidOperationException($"Cannot return a null member within a non-null list for list index {index}.");
+                                }
+                                else
+                                {
+                                    SetNodeError(context, node, new ExecutionError("Cannot return a null member within a non-null list."));
+                                }
```

Comment:
After #2341 this block of code may change a bit.

Reply:
> After #2341 this block of code may change a bit.

Yes it will chance significantly.  But this PR applies to the current state of the project.